jvb/src/main/kotlin/org/jitsi/videobridge/JvbLoadManager.kt

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.videobridge
+
+import org.jitsi.nlj.util.NEVER
+import org.jitsi.utils.logging2.cdebug
+import org.jitsi.utils.logging2.createLogger
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.util.logging.Level
+
+class JvbLoadManager<T : JvbLoadMeasurement> @JvmOverloads constructor(
+    private val jvbLoadThreshold: T,
+    private val loadReducer: JvbLoadReducer,
+    private val clock: Clock = Clock.systemUTC()
+) {
+    private val logger = createLogger(minLogLevel = Level.ALL)
+
+    private var lastReducerTime: Instant = NEVER
+
+    fun loadUpdate(loadMeasurement: T) {
+        logger.cdebug { "Got a load measurement of $loadMeasurement" }
+        if (loadMeasurement.getLoad() >= jvbLoadThreshold.getLoad()) {
+            if (Duration.between(lastReducerTime, clock.instant()) >= loadReducer.impactTime()) {
+                logger.info("Load measurement $loadMeasurement is above threshold of $jvbLoadThreshold, " +
+                        "running load reducer")
+                loadReducer.reduceLoad()
+                lastReducerTime = clock.instant()
+            } else {
+                logger.info("Load measurement $loadMeasurement is above threshold of $jvbLoadThreshold, " +
+                        "but load reducer started running ${Duration.between(lastReducerTime, clock.instant())} " +
+                        "ago, and we wait ${loadReducer.impactTime()} between runs")
+            }
+        }
+    }
+}
+
+interface JvbLoadMeasurement {
+    fun getLoad(): Double
+}
+
+class RtpPacketDelayMeasurement(private val rtpDelay: Double) : JvbLoadMeasurement {
+    override fun getLoad(): Double = rtpDelay
+
+    override fun toString(): String = "RTP packet delay of $rtpDelay ms"
+}
+
+val RtpPacketDelayThreshold = RtpPacketDelayMeasurement(30.0)
+
+interface JvbLoadReducer {
+    fun reduceLoad()
+    fun impactTime(): Duration
+}
+
+class LastNReducer(
+    private val videobridge: Videobridge,
+    private val reductionScale: Double
+) : JvbLoadReducer {
+    private val logger = createLogger()
+    override fun reduceLoad() {
+        // Find the highest number of endpoints sending video in a single conference: we need
+        // to set the last-n value to a size smaller than that
+        val largestConfSize = videobridge.conferences
+            .map { conf -> conf.endpoints.count(AbstractEndpoint::isSendingVideo) }
+            .max() ?: run {
+                    logger.info("No active conferences, can't reduce load by reducing last n")
+                    return
+                }
+
+        val newLastN = (largestConfSize * reductionScale).toInt()
```

Comment:
Can this actually increase the Last-N?

Reply:
You mean if `reductionScale` is > `1.0`?

Reply:
I'm thinking the case where the client has asked last-n 5 and then we set it to 10 to "reduce" the load? Or we have a `min()` call somewhere that I missed ?

Reply:
Ah! Yes that's a good point.  I wanted to make sure the JVB one took priority, but you're right that we wouldn't want to make it bigger than it currently was.  I think it's probably reasonable that the JVB limit should only make it smaller, so maybe rather than a JVB request always overriding a client request, instead we can just always take the minimum-requested-last-n (between JVB and client)--but I'll think on that some more.

Reply:
Ok, yeah I think always taking the smallest requested last-n value makes more sense here.  Nice catch.